Docker compose update

Aaron-Ritter opened the pull request:
This is the update for fusionauth-site docker, using the latest PostgreSQL release, switching to OpenSearch and adding a weekly dependabot check which addresses the two issues reported:

https://github.com/FusionAuth/fusionauth-issues/issues/2520
https://github.com/FusionAuth/fusionauth-issues/issues/2521

This PR requires https://github.com/sonderformat-llc/fusionauth-example-docker-compose to be copied to FusionAuth and https://github.com/FusionAuth/fusionauth-site/pull/2758 to be merged as its referencing the update of email templates.

This PR requires https://github.com/FusionAuth/fusionauth-containers/pull/96 to be merged as it's referencing pending updates.



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206275038333277

escii commented:
The Page "Switching Search Engines" (/docs/lifecycle/manage-users/search/switch-search-engines/) requires an update to emphasize "opensearch" over "elasticsearch".
Also, renaming the configuration setting "elasticsearch" within the fusionauth core to "searchengine" will help avoid favoring any specific product name. This change ensures a more neutral and inclusive approach to the search engine configuration, maybe you need to keep "elasticsearch" as an alias to maintain backward compatibility.

escii reviewed (approved):
No apparent issues spotted. Cool examples.

Aaron-Ritter commented:
> The Page "Switching Search Engines" (/docs/lifecycle/manage-users/search/switch-search-engines/) requires an update to emphasize "opensearch" over "elasticsearch". Also, renaming the configuration setting "elasticsearch" within the fusionauth core to "searchengine" will help avoid favoring any specific product name. This change ensures a more neutral and inclusive approach to the search engine configuration, maybe you need to keep "elasticsearch" as an alias to maintain backward compatibility.

@mooreds I agree with Mitch that it is confusing but it goes way beyond the scope of the change, I would suggest to create a issue to address this separately.

Beside that the PR is ready.

mooreds commented on astro/src/content/docs/_shared/_elasticsearch-ram.mdx:
```diff
@@ -1 +1 @@
-The RAM required by Elasticsearch depends on your login volume and user/entity counts. If you expect to have a few logins per minute and only a few thousand users/entities, `1GB-2GB` of RAM will be sufficient. If you have more users/entities as well as logins, please size up accordingly. We recommend running load tests of FusionAuth to help determine the amount of RAM that is required for your ne ...
+The RAM required by Elasticsearch and OpenSearch depends on your login volume and user/entity counts. If you expect to have a few logins per minute and only a few thousand users/entities, `1GB-2GB` of RAM will be sufficient. If you have more users/entities as well as logins, please size up accordingly. We recommend running load tests of FusionAuth to help determine the amount of RAM that is requi ...
```
```suggestion
The RAM required by Elasticsearch or OpenSearch depends on your login volume and user/entity counts. If you expect to have a few logins per minute and only a few thousand users/entities, `1GB-2GB` of RAM will be sufficient. If you have more users/entities as well as logins, please size up accordingly. We recommend running load tests of FusionAuth to help determine the amount of RAM that is required for your needs.
```

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -223,14 +128,22 @@ When running in development runtime mode, silent mode was enabled based upon the
 * Edit the Docker Compose file and specify a specific version. This is a good idea for a production deployment.
-* Remove the old image first: `docker rmi <old image id>`. This command may prompt you to remove containers using that image. Since all state is stored in the database, you can safely remove the containers. Then, when you restart Docker Compose, the latest image will be retrieved.
-* Pull the latest image with this command: `docker pull fusionauth/fusionauth-app:latest`.
+* Remove the old image first: `docker rmi <old image id>`. This command may prompt you to remove containers using that image. Since all state is stored in the database, you can safely remove the containers. Then, when you recreate Docker Compose, the latest image will be retrieved.
```
You aren't recreating docker compose, you are restarting docker compose which recreates the network, pulls images, etc. What am I missing?

Aaron-Ritter replied:
I completely redone the section, and describing any of it as restart doesn't make sense in general, unless we use docker compose stop/start/restart commands.

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -255,96 +168,13 @@ Here are the steps to do so:
+<RemoteCode title="Example docker-compose.yml for running Kickstart" url="https://raw.githubusercontent.com/sonderformat-llc/fusionauth-example-docker-compose/main/kickstart/docker-compose.yml" lang="yaml" />
 
-After running `docker compose up` you should see a line like this in the output:
+After running `docker compose up -d` you should see a line like this in the logs `docker compose logs fusionauth`:
```
```suggestion
After running `docker compose up -d` you should see a line similar to the one below in the logs. These logs can  be accessed using this command: `docker compose logs fusionauth`:
```

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -473,17 +219,23 @@ Such output indicates that the plugin has been installed and can be used.
 In this situation, do one of the following:
 
-* Use `host.docker.internal` as the hostname; this works on macOS and Windows hosts
+* Use `host.docker.internal` as the hostname instead of `localhost`; this works on macOS and Windows hosts, for linux you have to add `--add-host=host.docker.internal:host-gateway` to the docker engine.
```
```suggestion
* For macOS and Windows hosts, you can use `host.docker.internal` as the hostname instead of `localhost`.
* For Linux hosts, add `--add-host=host.docker.internal:host-gateway` to the docker engine.
```

Updated this for clarity, see next comment for correctness.

mooreds replied:
Can you explain what you mean by 'docker engine' here? Also, this post indicates you can do the same with the `extra_hosts` YAML key: https://stackoverflow.com/questions/70725881/what-is-the-equivalent-of-add-host-host-docker-internalhost-gateway-in-a-comp  

That seems preferable.

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -473,17 +219,23 @@ Such output indicates that the plugin has been installed and can be used.
+
+## OpenSearch Production Deployment Configuration
+
+The productive runtime requirements and configuration for OpenSearch will drastically differ from the docker compose examples as they're configured without any security and redundancy.
```
```suggestion
Production runtime requirements and configuration for OpenSearch will drastically differ from the docker compose examples as the examples are configured without any security or redundancy.
```

mooreds commented on astro/src/content/docs/lifecycle/manage-users/search/switch-search-engines.mdx:
```diff
@@ -10,7 +10,7 @@ import DataFieldDataTypeChanges from 'src/content/docs/_shared/_data-field-data-
 
 * `database`
-* `elasticsearch`
+* `elasticsearch` (for Elasticsearch and OpenSearch)
```
```suggestion
* `elasticsearch` (for Elasticsearch or OpenSearch)
```

mooreds reviewed (changes requested):
Please add to example apps; seems similar to example kickstarts repo: https://github.com/FusionAuth/fusionauth-site/blob/master/astro/src/content/json/exampleapps.json#L135

I moved over the repo. Please update everything to use https://github.com/fusionauth/fusionauth-example-docker-compose

I don't see where it depends on the fusionauth-containers repo. What am I missing?

Aaron-Ritter commented:
> I don't see where it depends on the fusionauth-containers repo. What am I missing?

The PR https://github.com/FusionAuth/fusionauth-containers/pull/96 changes the templates to use OpenSearch and are added in the first section of the documentation. Only once the PR is merged the description including OpenSearch in relation to the code makes sense again.

https://github.com/FusionAuth/fusionauth-site/blob/d67fd556ed47780c13e9d2a072fe18950548a1d5/astro/src/content/docs/get-started/download-and-install/docker.mdx#L24

https://github.com/FusionAuth/fusionauth-site/blob/d67fd556ed47780c13e9d2a072fe18950548a1d5/astro/src/content/docs/get-started/download-and-install/docker.mdx#L36

Aaron-Ritter commented:
@mooreds i went through and implemented your review feedback and discovered that there are again `docker-compose` commands instead of `docker compose` https://github.com/FusionAuth/fusionauth-site/pull/2310 which i changed once more too for the ones left over.

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -21,9 +21,9 @@ The following example is using Docker Compose. Here are [kubernetes installation
 All of the FusionAuth Docker images may be found on [Docker Hub](https://hub.docker.com/u/fusionauth/).
 
-The FusionAuth Docker Compose yaml files may be found on [FusionAuth containers repository](https://github.com/FusionAuth/fusionauth-containers).
+The FusionAuth Docker Compose yaml files may be found on [FusionAuth containers repository](https://github.com/FusionAuth/fusionauth-containers) and more examples in the [FusionAuth docker compose example repository](https://github.com/sonderformat-llc/fusionauth-example-docker-compose) described in more detail below.
```
You missed this reference to sonderformat-llc.

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -222,15 +127,25 @@ When running in development runtime mode, silent mode was enabled based upon the
-* Pull the latest image with this command: `docker pull fusionauth/fusionauth-app:latest`.
+* Pull the latest image with this command: `docker pull fusionauth/fusionauth-app:latest` or `docker compose pull` and recreate your deployment with `docker compose up -d` which will recreate every container where a new image is available.
+* Edit the Docker Compose file and specify a specific version. This is a good idea for a production deployment. With `docker compose pull` you can only pull the specified image, or just run `docker compose up -d` which will pull the image and recreate the container at the same time.
+* Remove all images with `docker rmi <IMAGE ID>` selected from the list of IMAGE IDs which you can show with `docker images`. This requires that the image isn't used and therefore may prompt you to remove containers using it. Since all state of FusionAuth is stored in the database, and for docker you use volumes to persist data, you can safely remove the containers. After that you can use `docker ...
```
I've never needed to use `docker compose build`. Does `docker compose up -d` do that implicitly?

Aaron-Ritter replied:
There are some scenarios where you use `docker compose build` especially in combination with `--pull` and `--no-cache` in scenarios where you use the latest tag. Or when you want to first test the build on its own after an version update.

mooreds replied:
Okay, that's fine, let's leave it in there then.

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -222,15 +127,25 @@ When running in development runtime mode, silent mode was enabled based upon the
+* Edit the Docker Compose file and specify a specific version. This is a good idea for a production deployment. With `docker compose pull` you can only pull the specified image, or just run `docker compose up -d` which will pull the image and recreate the container at the same time.
+* Remove all images with `docker rmi <IMAGE ID>` selected from the list of IMAGE IDs which you can show with `docker images`. This requires that the image isn't used and therefore may prompt you to remove containers using it. Since all state of FusionAuth is stored in the database, and for docker you use volumes to persist data, you can safely remove the containers. After that you can use `docker ...
+
+To generally clean up your images on your system `docker image prune` is useful too as it will remove all dangling images.
```
```suggestion
To generally clean up your images on your system ,`docker image prune` will remove all unused, dangling images.
```

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -255,96 +170,13 @@ Here are the steps to do so:
+<RemoteCode title="Example docker-compose.yml for running Kickstart" url="https://raw.githubusercontent.com/FusionAuth/fusionauth-example-docker-compose/main/kickstart/docker-compose.yml" lang="yaml" />
 
-After running `docker compose up` you should see a line like this in the output:
+After running `docker compose up -d` you should see a line similar to the one below in the logs. These logs can  be accessed using this command: `docker compose logs fusionauth`:
```
```suggestion
After running `docker compose up -d` you should see a line similar to the one below in the logs. These logs can be accessed using this command: `docker compose logs fusionauth`:
```

mooreds commented on astro/src/content/docs/get-started/download-and-install/docker.mdx:
```diff
@@ -366,103 +198,19 @@ docker compose down -v
+<RemoteCode title="Example docker-compose.yml for installing a plugin" url="https://raw.githubusercontent.com/FusionAuth/fusionauth-example-docker-compose/main/plugin/docker-compose.yml" lang="yaml" />
 
-After running `docker compose up` you should see a line like this in the output:
+After running `docker compose up -d` you should see a line like this in the logs `docker compose logs fusionauth`:
```
```suggestion
After running `docker compose up -d` you should see a line like this in the logs, which you can access using the command `docker compose logs fusionauth`:
```

vcampitelli reviewed (changes requested):
I've opened https://github.com/sonderformat-llc/fusionauth-site/pull/1 with a few changes

Aaron-Ritter commented:
merged review changes by @vcampitelli

mooreds commented:
@vcampitelli why did you ask for my review? Is it because you can't self review or something? Is this ready to ship?
